### PR TITLE
add shallow slice method

### PR DESCRIPTION
**Pull request by geloescht**
This PR adds the method shallowSlice which is just like slice but returns a BufferList instead of Buffer. It never copies anything but only slices buffers where necessary.

It also fixes a minor issue in _offset that resulted in undefined being returned when indexing past the end of the BufferList.

### CODE REVIEW

**Comment by mcollina** [comment]
LGTM

File: bl.js [hunk 1]
```diff
@@ -174,6 +174,31 @@ BufferList.prototype.copy = function copy (dst, dstStart, srcStart, srcEnd) {
   return dst
 }
 
+BufferList.prototype.shallowSlice = function shallowSlice (start, end) {
+  start = start || 0
+  end = end || this.length
+
+  if(start < 0)
```
Comment (rvagg) [inline bl.js]: whitespace between `if` and `(` pls

File: bl.js [hunk 2]
```diff
@@ -174,6 +174,31 @@ BufferList.prototype.copy = function copy (dst, dstStart, srcStart, srcEnd) {
...
+    start += this.length
+  if(end < 0)
+    end += this.length
+
+  var startOffset = this._offset(start)
+  var endOffset = this._offset(end)
```
Comment (rvagg) [inline bl.js]: no multiple `var` statements please, these 3 should be bunched up in one, comma-first, see `BufferList.prototype.copy` for a big example

File: bl.js [hunk 3]
```diff
@@ -174,6 +174,31 @@ BufferList.prototype.copy = function copy (dst, dstStart, srcStart, srcEnd) {
...
+
+  var buffers = this._bufs.slice(startOffset[0], endOffset[0]+1)
```
Comment (rvagg) [inline bl.js]: whitespace around `+` pls

**Comment by rvagg** [comment]
sgtm I guess, sans nits and also README addition

**Comment by geloescht** [comment]
Will fix the formatting and add the entry into the README next week.

**Comment by geloescht** [comment]
Please review.

**Comment by mcollina** [comment]
Can you please add a unit test that checks if it's shallow? meaning that changing the original would alter the copied one?

**Comment by geloescht** [comment]
done :)

File: bl.js [hunk 4]
```diff
@@ -40,7 +40,7 @@ BufferList.prototype._offset = function _offset (offset) {
   var tot = 0, i = 0, _t
   for (; i < this._bufs.length; i++) {
     _t = tot + this._bufs[i].length
-    if (offset < _t)
+    if (offset < _t || i == this._bufs.length - 1)
```
Comment (mcollina) [inline bl.js]: why this change?

Comment (geloescht) [inline bl.js]: Because _offset() returned undefined when an index >= length was requested. Most methods of Buffer deal with an index out of range gracefully. I think _offest() should not be concerned with range checking. If we return undefined, even the most simple assumption, that _offset() returns an array of two elements, is broken.

Comment (mcollina) [inline bl.js]: Perfect, thanks for the clarification!

**Comment by mcollina** [comment]
LGTM.

@rvagg, any opinions on this?
